File: src/main/java/generators/StructGenerator.java
package generators;

import io.swagger.models.HttpMethod;
import io.swagger.models.Model;
import io.swagger.models.Operation;
import io.swagger.models.Response;
import io.swagger.models.parameters.Parameter;
import io.swagger.models.properties.Property;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import util.FrugalTypeResolver;

import java.util.Map;

public class StructGenerator {

    private static Logger logger = LoggerFactory.getLogger(StructGenerator.class);

    public static String generate(String name, Model m) {
        StringBuilder builder = new StringBuilder();
        builder.append(generateHeader(name));
        int index = 1;
        for (String key : m.getProperties().keySet()) {
            Property prop = m.getProperties().get(key);
            builder.append(generateProperty(key, prop, index));
            index++;
        }
        builder.append(generateFooter());
        return builder.toString();
    }

    public static String generateRequest(Operation o) {
        String requestName = String.format("%sRequest", o.getOperationId());
        return generate(requestName, o);
    }

    public static String generate(String name, Operation o) {
        StringBuilder builder = new StringBuilder();
        builder.append(generateHeader(name));
        int index = 1;
        for (Parameter p : o.getParameters()) {
            builder.append(generateProperty(p.getName(), p, index));
            index++;
        }
        builder.append(generateFooter());
        return builder.toString();
    }

    private static String generateFooter() {
        return "}\n\n";
    }

    private static String generateProperty(String key,
                                           Property prop,
                                           int index) {
        return String.format("    %d: %s %s\n",
                             index,
                             FrugalTypeResolver.resolve(key, prop),
                             key);
    }

    private static String generateKnownProperty(String key,
                                                String prop,
                                                int index) {
        return String.format("    %d: %s %s\n", index, prop, key);
    }

    private static String generateProperty(String key,
                                           Parameter parameter,
                                           int index) {
        return String.format("    %d: %s %s\n",
                             index,
                             FrugalTypeResolver.resolve(key, parameter),
                             key);
    }

    private static String generateHeader(String name) {
        return String.format("struct %s {\n", capitalize(name));
    }

    public static String generateResponse(Operation o, HttpMethod method) {
        StringBuilder builder = new StringBuilder();
        String responseName = String.format("%sResponse", capitalize(o.getOperationId()));
        builder.append(generateHeader(responseName));
        builder.append(generateKnownProperty("code", "i32", 1));
        builder.append(generateKnownProperty("message", "string", 2));
        Map<String, Response> responses = o.getResponses();
        int index = 3;
        // GET requests will have a custom output on success
        if (responses.get("200") != null && method == HttpMethod.GET) {
            Property p = responses.get("200").getSchema();
            String name = p.getName() == null ? "items" : p.getName();
            builder.append(generateProperty(name, p, index));
        } else {
            // Return what was sent
            for (Parameter p : o.getParameters()) {
                builder.append(generateProperty(p.getName(), p, index));
                index++;
            }
        }
        builder.append(generateFooter());
        return builder.toString();
    }

    private static String capitalize(String in) {
        return in.substring(0, 1).toUpperCase() + in.substring(1);
    }
}

File: generators/struct_generator.py
from utils.frugal_type_resolver import FrugalTypeResolver
from utils.logging import get_logger

_logger = get_logger('generators.StructGenerator')


class StructGenerator(object):

    @staticmethod
    def generate(name: str, attributes: dict) -> str:
        _logger.debug(f'Generating struct for `{name}`')
        struct = StructGenerator._generate_header(name)
        # There are multiple ways for a struct to get properties:
        #   - the `parameters` tag
        #   - referenced using the `allOf` tag
        #   - the `properties` tag
        combined_properties = {}
        parameters: list = attributes.get('parameters', [])
        for parameter in parameters:
            combined_properties[parameter['name']] = parameter
        all_of: list = attributes.get('allOf', [])
        for propMap in all_of:
            combined_properties.update(propMap['properties'])
        properties: dict = attributes.get('properties', {})
        combined_properties.update(properties)

        assert combined_properties, f'Missing properties for {name}'
        index = 0
        for name, spec in combined_properties.items():
            index += 1
            struct += StructGenerator._generate_property(name, spec, index)
        struct += StructGenerator._generate_footer()
        return struct

    @staticmethod
    def _generate_header(name: str) -> str:
        return f'struct {name} {{\n'

    @staticmethod
    def _generate_footer() -> str:
        return '}\n'

    @staticmethod
    def _generate_property(name: str, prop: dict, index: int) -> str:
        return f'    {index}: {FrugalTypeResolver.resolve(name, prop)} {name}\n'

    @staticmethod
    def generate_request(operation: dict):
        request_name = operation.get('operationId')
        assert request_name, 'Missing operationId'
        return StructGenerator.generate(f'{request_name}Request', operation)
